components/script/task_queue.rs

```diff
@@ -53,22 +58,85 @@ impl<T: QueuedTaskConversion> TaskQueue<T> {
             msg_queue: DomRefCell::new(VecDeque::new()),
             taken_task_counter: Default::default(),
             throttled: Default::default(),
+            inactive: Default::default(),
+        }
+    }
+
+    /// Process an incoming message, either:
+    /// 1: making it available for further processing as part of "incoming", or
+    /// 2: in the case of a not fully-active document, storing it into the "inactive" queue.
+    fn process_message(
+        &self,
+        msg: T,
+        currently_inactive: &HashSet<PipelineId>,
+        incoming: &mut Vec<T>,
+    ) {
+        if msg.is_wake_up() {
+            return;
+        }
+        let mut inactive = self.inactive.borrow_mut();
+        // Hold back tasks for currently not fully-active documents.
+        // https://html.spec.whatwg.org/multipage/#event-loop-processing-model:fully-active
+        if let Some(pipeline_id) = msg.pipeline_id() {
+            if currently_inactive.contains(pipeline_id) {
+                let inactive_queue = inactive
+                    .entry(pipeline_id.clone())
+                    .or_insert(VecDeque::new());
+                inactive_queue
+                    .push_back(msg.into_queued_task().expect(
+                        "Incoming messages should always be convertible into queued tasks",
+                    ));
+                if incoming.is_empty() {
+                    // Ensure there is at least one message.
+                    // Otherwise if the just stored inactive message
+                    // was the first and last of this iteration,
+                    // it will result in a spurious wake-up of the event-loop.
+                    incoming.push(T::inactive_msg());
+                }
+                return;
+            }
+        }
+        incoming.push(msg);
+    }
+
+    /// Release previously held-back tasks for documents that are now fully-active.
+    fn release_tasks_for_fully_active_documents(
+        &self,
+        currently_inactive: &HashSet<PipelineId>,
+        incoming: &mut Vec<T>,
+    ) {
+        let previously_inactive: HashSet<PipelineId> = {
+            let inactive = self.inactive.borrow();
+            inactive.keys().map(|id| id.clone()).collect()
+        };
+        let mut inactive = self.inactive.borrow_mut();
+        for pipeline_id in previously_inactive.difference(&currently_inactive) {
```

Comment:
AFAICT, there is no need to collect the inactive IDs, just loop over the entries of `self.inactive` and check that the key isn't in `currently_inactive`.